File: CS/EFCore/ASP.NET Core/Blazor/RuntimeDbChooser.Blazor.Server/BlazorApplication.cs
using System.Collections.Concurrent;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Blazor;
using Microsoft.Extensions.DependencyInjection;
using RuntimeDbChooser.Blazor.Server.Services;
using RuntimeDbChooser.Services;

namespace RuntimeDbChooser.Blazor.Server {
    public class RuntimeDbChooserBlazorApplication : BlazorApplication {
        private static ConcurrentDictionary<string, bool> isCompatibilityChecked = new ConcurrentDictionary<string, bool>();

        protected override bool IsCompatibilityChecked {
            get => isCompatibilityChecked.ContainsKey(ServiceProvider.GetRequiredService<IConnectionStringProvider>().GetConnectionString());

            set => isCompatibilityChecked.TryAdd(ServiceProvider.GetRequiredService<IConnectionStringProvider>().GetConnectionString(), value);
        }

        protected override void OnSetupStarted() {
            base.OnSetupStarted();
#if DEBUG
            if(System.Diagnostics.Debugger.IsAttached && CheckCompatibilityType == CheckCompatibilityType.DatabaseSchema) {
                DatabaseUpdateMode = DatabaseUpdateMode.UpdateDatabaseAlways;
            }
#endif
#if EASYTEST
            DatabaseUpdateMode = DatabaseUpdateMode.UpdateDatabaseAlways;
#endif
        }
        protected override LogonController CreateLogonController() {
            return CreateController<Controllers.CustomLogonController>();
        }
    }
}

File: VB/XPO/ASP.NETCore/Blazor/RuntimeDbChooser.Blazor.Server/BlazorApplication.vb
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Blazor
Imports Microsoft.Extensions.DependencyInjection
Imports RuntimeDbChooser.Services
Imports System.Collections.Concurrent

Namespace RuntimeDbChooser.Blazor.Server

    Public Class RuntimeDbChooserBlazorApplication
        Inherits BlazorApplication

        Private Shared isCompatibilityCheckedField As ConcurrentDictionary(Of String, Boolean) = New ConcurrentDictionary(Of String, Boolean)()

        Protected Overrides Property IsCompatibilityChecked As Boolean
            Get
                Return isCompatibilityCheckedField.ContainsKey(ServiceProvider.GetRequiredService(Of RuntimeDbChooser.Services.IConnectionStringProvider)().GetConnectionString())
            End Get

            Set(ByVal value As Boolean)
                isCompatibilityCheckedField.TryAdd(ServiceProvider.GetRequiredService(Of RuntimeDbChooser.Services.IConnectionStringProvider)().GetConnectionString(), value)
            End Set
        End Property

        Protected Overrides Sub OnSetupStarted()
            MyBase.OnSetupStarted()
#If DEBUG
            If System.Diagnostics.Debugger.IsAttached AndAlso CheckCompatibilityType Is CheckCompatibilityType.DatabaseSchema Then
                DatabaseUpdateMode = DatabaseUpdateMode.UpdateDatabaseAlways
            End If
#End If
        End Sub

        Protected Overrides Function CreateLogonController() As LogonController
            Return CreateController(Of Controllers.CustomLogonController)()
        End Function
    End Class
End Namespace
